[FLINK-21159][connector/kafka] Send NoMoreSplitsEvent to all readers even if the reader is not assigned with any splits

PatrickRen opened the pull request:
## What is the purpose of the change

This pull request fixes bug in ```KafkaSourceEnumerator``` that only sending ```NoMoreSplitsEvent``` to readers assigned with splits, which will lead to job hanging forever even if the source is bounded.

## Brief change log

*(for example:)*
  - Signal ```NoMoreSplitsEvent``` to reader in ```addReader``` and ```addSplitsBack``` instead of during split assigning
  - Add IT case for testing cases that a reader is not assigned with any splits
  - Fix a typo in JavaDoc of SplitEnumeratorContext#callAsync

## Verifying this change
This change added tests and can be verified as follows:
  - *Add IT case for testing cases that a reader is not assigned with any splits*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

PatrickRen commented:
I noticed a race condition in my modification: the variable ```noMoreNewPartitionSplits``` may not be set when ```addReader``` is invoked because they are in different threads, which will cause not sending NoMoreSplitsEvent to some readers. I'll make another commit later.

becketqin commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java:
```diff
@@ -64,6 +64,8 @@
     private final Properties properties;
     private final long partitionDiscoveryIntervalMs;
     private final SplitEnumeratorContext<KafkaPartitionSplit> context;
+    private volatile boolean partitionDiscoveryFinished = false;
```
It seems that we can just reuse the `noMoreNewPartitionSplits` flag.

- If the periodic partition discovery is disabled, then after the first partition discovery is done, set the `noMoreNewPartitionSplits` flag to true. The subsequent `assignPendingPartitionSplits` will just send the `NoMoreSplitsEvent` to all the readers. As long as the flag is set in the main thread, the readers who registered before the first partition discovery is done will not receive duplicate `NoMoreSplitsEvent` in this case.
- Otherwise, the `noMoreNewPartitionSplits` will always be set to false, and no `NoMoreSplitsEvent` will be sent.

becketqin commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumerator.java:
```diff
@@ -276,16 +284,21 @@ private void assignPendingPartitionSplits() {
+                        (readerId, ignore) -> {
+                            if (!finishedReaders.contains(readerId)) {
+                                context.signalNoMoreSplits(readerId);
+                                finishedReaders.add(readerId);
```
What happens if the readers failover? Do we assume that the source readers remember the reception of `NoMoreSplitsEvent`?

becketqin reviewed (commented):
@PatrickRen Thanks for the patch. I left a few comments.

PatrickRen commented:
@becketqin Thank you~ I changed my implementation and made another commit. 

Now ```NoMoreSplitsEvent``` will be sent in ```addReader()``` so all readers will be notified and we don't need to "remember" readers that have been signaled. 

There is possibility that ```addReader()``` could be invoked before the first partition discovery and split assignment, so I add a set called ```finishingReaders``` for tracking these early-added readers, and signal ```MoMoreSplitsEvent``` to them later after the first discovery and assignment.

PatrickRen commented:
Already fixed by dc3d0702087c8681435761eed7f2aa9161b09540.
